### ISSUE TITLE
Allow passing `type` property to AST constructors

### ISSUE THREAD

**Issue by RReverser** [shapesecurity/shift-ast-js#49]
I think it's worth allowing to pass [an optional and valid, but useless] `type` property to the constructors, so that you could take a node instance and perform a shallow copy by passing it to its own constructor.

E.g. currently `checked.js` doesn't allow this, but I think it's harmless to support and sometimes useful to do:

```js
> a = new AST.LiteralNumericExpression({value:42})
LiteralNumericExpression { type: 'LiteralNumericExpression', value: 42 }
> b = new AST.LiteralNumericExpression(a)
Thrown:
TypeError: Argument to LiteralNumericExpression constructor has wrong keys: expected {value}, got {type, value}
    at new LiteralNumericExpression (C:\Users\Ingvar\shift-ast-js\dist\checked.js:1405:11)
```

This could be particularly useful when constructing Shift node instances from POJO objects.

**Comment by bakkot** [comment]
I guess I'm fine with it, but I'd probably want to assert that the `type` parameter, if passed, held the name of the type the constructor is intending to construct. The `checked` constructors are mainly intended for use in development and debugging, so they are intended to be extremely strict.

**Comment by RReverser** [comment]
Yeah, absolutely, assertion is what I meant by "optional and valid, but useless".
